lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,198 @@
+    #   },
+    #   "123.11.33.0/24":{
+    #      "00:0a:95:9d:68:33": {"ip": "123.11.33.1", "timestamp": "2019-09-17 12:04:43 -0400"}
+    #      "00:0a:95:9d:68:34": {"ip": "123.11.33.2", "timestamp": "2019-09-17 12:05:48 -0400"}
+    #      "00:0a:95:9d:68:35": {"ip": "123.11.33.3", "timestamp:: "2019-09-17 12:06:50 -0400"}
+    #   }
+    # }
+    #
+    # Params: 1. subnet:  The Foreman subnet object
+    #         2. mac:     The mac address of the interface obtaining the IP address for
+    #
+    # Returns: A hash with next available IP in the "data" field, or a hash with "error" on failure
+    #
+    # Examples:
+    #   Response if success:
+    #     {"code": 200, "success": true, "data": "100.55.55.3", "time": 0.012}
+    #   Response if missing required params:
+    #     {"error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","A 'mac' address must be provided(e.g. 00:0a:95:9d:68:10)"]}
+    #   Response if subnet error:
+    #     {"error": "The specified subnet does not exist in External IPAM."}
+    #   Response if there are no free addresses:
+    #     {"error": "No free addresses found"}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def next_ip(subnet, mac)
+      raise "subnet.network cannot be nil" unless subnet.network.present?
+      parse get("/next_ip?cidr=#{subnet.network_address}&mac=#{mac}")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to retrieve the next available IP for subnet %s from External IPAM."), subnet.network_address)
+    end
+
+    # Adds an IP address to the specified subnet in External IPAM. This will reserve the IP in the
+    # External IPAM database.
+    #
+    # Params: 1. ip:     IP address to be added
+    #         2. subnet: The Foreman subnet object
+    #
+    # Returns: A hash with a message of "Address created" on success, or a hash with "error" on failure
+    #
+    # Examples:
+    #   Response if success:
+    #     {"code": 201, "success": true, "message": "Address created", "id": "156", "time": 0.015}
+    #   Response if IP already reserved:
+    #     {"code": 409, "success": false, "message": "IP address already exists", "time": 0.006}
+    #   Response if subnet error:
+    #     {"error": "The specified subnet does not exist in External IPAM."}
+    #   Response if missing required params:
+    #     {"error": ["A 'cidr' parameter for the subnet must be provided(e.g. 100.10.10.0/24)","Missing 'ip' parameter. An IPv4 address must be provided(e.g. 100.10.10.22)"]}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def add_ip_to_subnet(ip, subnet)
+      raise "subnet.network cannot be nil" unless subnet.network.present?
+      parse(post({:ip => ip, :cidr => subnet.network_address}, "/add_ip_to_subnet"))
```

Comment:
I've seen `success?` helper method somewhere above, I suggest to move the error checking code here and raise ProxyException or Foreman::Exception in case it's not success. Otherwise return the valid Ruby hash so the caller simply can expect valid result. That's a sign of a good contract.